developer_tools/minprintf.h

```diff
@@ -0,0 +1,336 @@
+ * @param width Similar to width parameter of print_int_value().
+ * @param precision Tells number of digits to be printed after decimal (By default it is 8). 
+ */
+void print_double_value(double dval, int width, int precision)
+{
+	int ndigits = get_number_of_digits((int) dval); // Store number of digits before decimal in dval.
+	int reqd_blanks = width - (precision + 1) - ndigits; // Blanks to be printed before printing dval, just to cover the width.
+	
+	print_int_value((int) dval, reqd_blanks, 0); // Prints the part before decimal.
+
+	put_char('.'); // Print decimal.
+
+	/*Deletes digits before decimal and makes them zero. For example:
+		if dval = 1923.79022, them this will make dval = 0.79022.
+	*/
+	dval = dval - (int) dval;
+
+	dval *= power_of_ten(precision); // Brings precision number of digits after decimal to before decimal.
+	
+	print_int_value((int) dval, 0, precision); // Prints the remaining number.
+}
+
+/**
+ * First size of string is calculated to check weather
+ * width and precision is to be taken into account or not.
+ * Then string is printed in accordingly.
+ * @param p pointer to string to be printed.
+ * @param width if (width > sizeof string) then blanks will be printed before sting to cover up width.
+ * @param precision Total charanters of string to be printed (Prints whole string if 0 or greater than size of string).
+*/
+void print_string(char *p, int width, int precision)
+{
+	int size = 0; // Stores number of character in string.
+	char *s = p; // Temporary pointer.
+
+	/* Calculates size of string p*/
+	while (*s != '\0') { 
+		++size;
+		++s;
+	}
+
+	s = p; // Point s to starting of p.
+
+	/* Checks how many characters to be printed.
+		if precision is defined then size variable is changed to precision so that only precision
+		number of characters were printed.
+	*/
+	if (precision != 0 && precision < size)
+		size = precision;
+
+	/* Prints blanks to cover the width if required.*/
+	for (int i = 0; i < (width - size); ++i)
+		put_char(' ');
+
+	/* Print the string.*/
+	for (int i = 0; i < size; ++i)
+		put_char(*s++);
+
+}
+
+/**
+ * Takes width and precision specified from format of string.
+ * @param p pointer of format string.
+ * @param width variable in which width will be stored.
+ * @param precision variable in which precision will be stored.
+ * @return character pointer to the current pointer of string p. (Used to update vaue of p)
+*/
+char *get_width_and_precision(char *p, int *width, int *precision)
+{
+	/* Skip % if p is pointing to it.*/
+	if (*p == '%')
+		++p;
+	
+	/* Calculates the width specified. */
+	while (*p != '.' && is_number(p)) 
+		*width = *width * 10 + (*p++ - '0');
+
+	/* Calculates the precision specified.*/
+	if (*p == '.' /* Since a precision is always specified after a '.'. */) { 
+		while (is_number(++p))
+			*precision = *precision * 10 + (*p - '0'); 
+		unget_ch(p); // The non number is stored in a buffer.
+	}
+	return p;
+}
+
+/**
+ * minprintf is the main function.
```

Comment:
The main function is **always** called `main` and is of `int` value (see the [typical structure](https://github.com/TheAlgorithms/C/blob/master/CONTRIBUTING.md#typical-structure-of-a-program) of a program for more information).